crates/goose-llm/src/message.rs

```diff
@@ -0,0 +1,616 @@
+use std::{collections::HashSet, iter::FromIterator, ops::Deref};
+
+/// Messages which represent the content sent back and forth to LLM provider
+///
+/// We use these messages in the agent code, and interfaces which interact with
+/// the agent. That let's us reuse message histories across different interfaces.
+///
+/// The content of the messages uses MCP types to avoid additional conversions
+/// when interacting with MCP servers.
+use chrono::Utc;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use smallvec::SmallVec;
+
+use crate::types::core::{Content, ImageContent, Role, TextContent, ToolCall, ToolResult};
+
+mod tool_result_serde;
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ToolRequest {
+    pub id: String,
+    #[serde(with = "tool_result_serde")]
+    pub tool_call: ToolResult<ToolCall>,
+}
+
+impl ToolRequest {
+    pub fn to_readable_string(&self) -> String {
+        match &self.tool_call {
+            Ok(tool_call) => {
+                format!(
+                    "Tool: {}, Args: {}",
+                    tool_call.name,
+                    serde_json::to_string_pretty(&tool_call.arguments)
+                        .unwrap_or_else(|_| "<<invalid json>>".to_string())
+                )
+            }
+            Err(e) => format!("Invalid tool call: {}", e),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ToolResponse {
+    pub id: String,
+    #[serde(with = "tool_result_serde")]
+    pub tool_result: ToolResult<Vec<Content>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ToolConfirmationRequest {
+    pub id: String,
+    pub tool_name: String,
+    pub arguments: Value,
+    pub prompt: Option<String>,
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct ThinkingContent {
+    pub thinking: String,
+    pub signature: String,
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct RedactedThinkingContent {
+    pub data: String,
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FrontendToolRequest {
+    pub id: String,
+    #[serde(with = "tool_result_serde")]
+    pub tool_call: ToolResult<ToolCall>,
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct ContextLengthExceeded {
+    pub msg: String,
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+/// Content passed inside a message, which can be both simple content and tool content
+#[serde(tag = "type", rename_all = "camelCase")]
+pub enum MessageContent {
+    Text(TextContent),
+    Image(ImageContent),
+    ToolRequest(ToolRequest),
+    ToolResponse(ToolResponse),
+    ToolConfirmationRequest(ToolConfirmationRequest),
+    FrontendToolRequest(FrontendToolRequest),
```

Comment:
see other nit, i kind of think we get rid of this variant

Reply:
+1 i thought of this one too but wasn't sure since its in service design